torch/_inductor/ir.py

```diff
@@ -6684,6 +6684,220 @@ def codegen(self, wrapper):
         wrapper.codegen_while_loop(self)
 
 
+class SequentialScan(ExternKernel):
```

Comment:
It would be great if we could refactor this to use while_loop. Ideally we would not have multiple, separate codegens for loops in the ir.

Reply:
@zou3519 and also had a discussion on this: we feel we can decouple the lowering of scan and while_loop:

1. the SequentialScan ir node scan carries clearer information (e.g. the number of iterations) compared with while_loop, where we have to inspect the cond_fn to infer this information. So at least we want to have a dedicated IR node for it.
2. We could still share code in wrapper codegen between scan and while_loop when necessary.
3. It's not that much work and it allows us to change them separately (e.g. adding new kwargs) without worrying about composability. After all, we only have a few of these so the complexity is managable.

Reply:
I understand having a separate IR node higher in the compile process, but I would expect them to desugar to the same IR in inductor lowering (for instance, maybe through replace_by_example).

Reply:
em...I think I need a bit clarification. Is the plan that when lowering scan, we create the out variant graph via re-tracing (i.e. what replace_by_example is actually doing behind the scene). Then we create a while_loop IR node instead?

Reply:
Btw, the thing about scan is that you know how long it's going to execute for. A while_loop can execute for a data-dependent amount of iterations. Would this change how the lowering works?

Reply:
One example is: you can memory plan a scan into having a scan_out variant. You can't memory plan a data-dependent while-loop (since you don't know how many iterations it's going to run)!

Reply:
> A while_loop can execute for a data-dependent amount of iterations. 

I guess we can still use a while True + break as long as we capture the dynamism inside the cond function.

>  You can't memory plan a data-dependent while-loop
This is not an issue for inference since we can assume while_loop's inputs and outputs has same meta in every iteration, we can probably create the output buffer based on the input. 

When we consider autograd for while_loop, in each iteration, we want to checkpoint some intermediate outputs for backward to use. During lowering, these intermediate outputs cannot be memory planned as of right now because we're not sure how much memory it's gonna use. We need some special ways of allocating the buffers maybe in wrapper codegen. But I guess we can still merge in the while True + break wrapper template under some flag.

Reply:
>  while_loop can execute for a data-dependent amount of iterations. Would this change how the lowering works?

At least with TS, we had [prim::Loop](https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/OVERVIEW.md#loops) which had a max trip count, an initial condition, and a loop carried condition. We could do something like this but we don't have to replicate exactly. We could have a class ForLoop that extends from WhileLoop and overrides cond or something. The important thing is that most of the logic is shared.

> In scan we need the symbolic index to index into the write positions but while_loop doesn't need to use that index. The consequence is that there maybe a dangling symints floating around if we don't have flag to control it.

The desugaring of the symint can happen prior to inductor graph lowering. I dont think this needs to affect the WhileLoop ir. 

> The reason I ask is that it seems we don't need a pattern matcher because we could match scan exactly during lowering

Yea, this is true. A similar pass that always decomposes operators and be used as inspiration is [here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/post_grad.py#L807-L815).

Reply:
updated with a while_loop lowering. I struggled at how to get construct the outputs of the while_loop so that inductor knows the output is some mutated inputs. Left 2 TODOs on this. Can I have some guidance on this? @Chillee @eellison

Reply:
Haven't checked how cpp wrapper goes. After fixing the TODOs, i'll add the corresponding tests and if the change is small to enable cpp wrapper, i can do i in this PR.

Reply:
Addressed the TODOs by adding the inputs to while_loop.mutation_outputs.